### PR TITLE
Implement searching for an existing family by parent name

### PR DESCRIPTION
**Pull request by jossmurphy**
### Notion ticket link

[Search for an existing client by name](https://www.notion.so/uwblueprintexecs/Search-for-an-existing-client-by-name-c9abac8344804d63ba8a9b3ca5f83ef4)

### Implementation description
- added the family search endpoint & types to FamilyAPI.ts
- created components to allow users to enter and search by parent name (StudentSearchBar) and retrieve a table of results (FamilySearchResultsTable)
  - these components will be reused for the "Add a guest" modal later, which is functionally & visually quite similar
- unit tests will be added in an upcoming PR

Screen Shot 2021-06-25 at 10 33 58 PM
Screen Shot 2021-06-25 at 10 34 17 PM

### Steps to test

1. open the deploy preview
2. go to the sessions page and click "Add a client"
3. search by the first and/or last name of a parent in the db, and verify that results appear
4. search by a first/last name that **does not** exist in the db, and verify that the table displays "No results found"
5. search with no first/last name, and verify that the input boxes highlight with a message "Please enter a value"
6. when leaving the search page (closing dialog, going to reg form, etc.) and returning, the search form & results should be a blank slate

### What should reviewers focus on?

- code quality
- manual testing functionality

### Checklist

- [X] My PR name is descriptive and in imperative tense
- [x] I have run the linter
- [X] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by brittanylau** [comment]
@jossmurphy nice work on this!! here's a quick summary of the additions I made:
* refactored out the search bar & results table to their own components (since we'll reuse this later for the "Add a guest" feature)
* used a `useEffect` hook to reset the form values when returning to the search form
* styling

feel free to take a look and let me know if I missed anything, or if you have any suggestions :)

File: src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      >
+        <Table stickyHeader aria-label="family search results table">
+          <caption hidden>Family search results</caption>
+          <TableHead>
+            <TableRow>
+              <TableCell>{DefaultFields.FIRST_NAME.name}</TableCell>
+              <TableCell>{DefaultFields.LAST_NAME.name}</TableCell>
+              <TableCell>{DefaultFields.PHONE_NUMBER.name}</TableCell>
+              <TableCell>{DefaultFields.EMAIL.name}</TableCell>
+              <TableCell>{DefaultFields.NUM_CHILDREN.name}</TableCell>
```
Comment (Tayef-Shah) [inline src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx:47]: Maybe we can replace this with something like
```suggestion
              {DefaultFamilySearchFields.map((searchField) => (
                <TableCell>{searchField.name}</TableCell>
              ))}
```

File: src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          </TableHead>
+          {families.length ? (
+            <TableBody>
+              {families.map((family) => (
+                <TableRow>
+                  <TableCell>{family.first_name}</TableCell>
+                  <TableCell>{family.last_name}</TableCell>
+                  <TableCell>{family.phone_number}</TableCell>
+                  <TableCell>{family.email}</TableCell>
+                  <TableCell>{family.num_children}</TableCell>
```
Comment (Tayef-Shah) [inline src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx:58]: I'm thinking we can maybe make a change similar to this
```suggestion
                  {DefaultFamilySearchFields.map((searchField) => (
                    <TableCell>{(family as any)[searchField.id]}</TableCell>
                  ))}
```

File: src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  TableHead,
+  TableBody,
+  TableRow,
+  TableCell,
+  TableFooter,
+  TableContainer,
+} from "@material-ui/core";
+import { makeStyles } from "@material-ui/styles";
+import { FamilySearchResponse } from "../../../api/FamilyAPI";
+import { DefaultFields } from "../../../constants/DefaultFields";
```
Comment (Tayef-Shah) [inline src/components/family-search/family-search-results-table/FamilySearchResultsTable.tsx:15]: Maybe let's make a list in DefaultFields for the search
```suggestion
import { DefaultFamilySearchFields } from "../../../constants/DefaultFields";
```
inside `DefaultFields` you'll need to make a list like
```
export const DefaultFamilySearchFields: DefaultField[] = [
  DefaultFields.FIRST_NAME,
  DefaultFields.LAST_NAME,
  DefaultFields.PHONE_NUMBER,
  DefaultFields.EMAIL,
  DefaultFields.NUM_CHILDREN,
];
```

**Review by Tayef-Shah — CHANGES_REQUESTED** [review]
Mapping the search fields may be cleaner in the long run. You'd just need to make a list of search fields in `DefaultFields.ts`. See below for details.

**Comment by brittanylau** [comment]
@Tayef-Shah I personally don't think this level of abstraction is necessary for this component - we should avoid casting to `any` to promote type safety & readability. it should only really be used to pacify errors [while migrating a js project to ts](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any), which is why it's in the sidebar component atm

we should probably be changing that instead – made [this ticket](https://www.notion.so/uwblueprintexecs/Developer-Hub-30e9b6c82af24054b31acf7c5e822c89?p=b6d63545230e48e88535bf84fd2a4006) to track that :)

**Comment by Tayef-Shah** [comment]
@brittanylau ah ok, sounds good then. I'll approve the changes.

**Review by Tayef-Shah — APPROVED** [review]

**Comment by brittanylau** [comment]
@Tayef-Shah thank you!! 🙌

**Pull request merged**